### PR TITLE
feat: cpu and store gas updates

### PR DESCRIPTION
**Pull request by piux2**
## Summary:

This PR updates the CPU and Store gas based on results from the benchmarking tool:https://github.com/gnolang/gno/pull/2241

For CPU gas, the measurement is in nanoseconds per opcode execution.
For storage gas, the measurement is in nanoseconds per byte for each type of Gno store access.

## Changes: 

We moved the gas meter from the underlying store to the upper Gno store to capture accurate resource consumption for VM transactions. At the same time, we retain the original gas store and gas meter for the Auth Keeper to handle regular blockchain transactions that do not necessarily involve the VM.

We also updated the gas-wanted in the integration test to reflect actual gas usage. This can serve as a flag to alert us to future changes that might increase gas assumptions.

Additional reasons for these changes include:

 - The Gno VM store processes additional complex encoding and decoding of data structures, rather than simply reading and writing bytes to disk.
 - For the above reason, we benchmarked gas for store access at the Gno store level.
 - We want to avoid consuming gas at two points for a single store access during a VM transaction.
 
Here are the diagrams to explain the store access gas before and after changes 

#### Before:

image

#### After:
image

<details><summary>Contributors' checklist...</summary>

- [x] Added new tests
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
</details>

### CODE REVIEW

File: gno.land/cmd/gnoland/testdata/addpkg_namespace.txtar
Comment (thehowl) [inline gno.land/cmd/gnoland/testdata/addpkg_namespace.txtar:1]: can we just make the txtars all just use a `-gas-wanted` of `10M`?

just because they'll have to be updated again. I'd want for us to keep track of "golden" gas values just in one place.

File: gnovm/pkg/gnolang/machine.go
Comment (thehowl) [inline gnovm/pkg/gnolang/machine.go:1157]: Can we add a reference to the benchmarks and methodology?

File: gno.land/pkg/sdk/vm/keeper.go
Comment (thehowl) [inline gno.land/pkg/sdk/vm/keeper.go:208]: This is unsafe, because vm.gnoStore is shared, and this would set the gas meter globally; ie. not scoped to the transaction.

File: gnovm/pkg/gnolang/store.go
Comment (thehowl) [inline gnovm/pkg/gnolang/store.go:764]: I think this ought to go in `BeginTransaction`, and the store shouldn't use a gas store otherwise.

File: gnovm/pkg/gnolang/store.go
Comment (thehowl) [inline gnovm/pkg/gnolang/store.go:127]: These are constants for the VM. Can these also not be constants?

File: tm2/pkg/sdk/context.go
Comment (thehowl) [inline tm2/pkg/sdk/context.go:181]: I can understand the rename to GasStore, but I don't think we should expose a gas-less `Store` method, especially when getting the underlying store is still trivial and exported.

**Review by zivkovicmilos — CHANGES_REQUESTED** [review]
I think the PR looks fine, apart from the comments left by @thehowl, which should be addressed about the global gas meter setting in the VM keeper 🙏 

I'm temporarily just putting a pause on the PR, to allow for us to discuss whether these new gas values make sense, since they are **a huge change**. I also want to see some more reasoning and gas comparisons, apart from https://github.com/gnolang/gno/pull/2241

If we're fine with the bumps, I am also cool with the PR moving forward as is ✅ 

cc @moul @notJoon

File: gnovm/pkg/gnolang/machine.go
Comment (moul) [reply gnovm/pkg/gnolang/machine.go:1157]: I'm concerned that this approach renders `ugnot` nearly valueless, as it results in excessively long and difficult-to-read numbers. 

We should either adopt a strategy that sacrifices some precision, such as dividing by N (which could be a parameter), or consider introducing another level of `gnot` conversion, like `ngnot` for gas, to maintain the relevance of `ugnot`.

File: gnovm/pkg/gnolang/store.go
Comment (moul) [reply gnovm/pkg/gnolang/store.go:764]: It makes sense to count gas for the store independently since future features will dissociate CPU gas from storage gas, allowing costs to cover `cpuCost + storageCost + futureStorageRent`. Additionally, realms could have different storage cost ratios—some sponsored by the chain, others transitioned to 'archives,' or even storage classes with varying configurations based on price tiers.

File: tm2/pkg/sdk/context.go
Comment (moul) [reply tm2/pkg/sdk/context.go:181]:
```suggestion
func (c Context) GaslessStore(key store.StoreKey) store.Store {
	return c.MultiStore().GetStore(key)
}

```

?

**Review by moul — APPROVED** [review]
Overall, it looks good. In this PR or a future one, I suggest we:  
-  Add a global multiplier/divider parameter to avoid using `ngnot` while keeping the numbers relatively readable.  
-  Create a benchmarking tool that runs all the gnocode in the repository and generates a CSV or TXT file with the gas details we commit periodically (there's no need for this on every commit, in my opinion). This way, whenever we want to fine-tune a parameter, we can run the same script again and compare the results.

**Review by zivkovicmilos — APPROVED** [review]
Unblocking after @moul is fine with the bumps

File: gno.land/cmd/gnoland/testdata/addpkg_namespace.txtar
Comment (piux2) [reply gno.land/cmd/gnoland/testdata/addpkg_namespace.txtar:1]: I understand that we need to update for gas changes. This gives us an opportunity to verify the changes, especially if there is a significant gas increase in our implementation. It acts as a safeguard. I am also open to other solutions.

File: gnovm/pkg/gnolang/machine.go
Comment (piux2) [reply gnovm/pkg/gnolang/machine.go:1157]: @moul @zivkovicmilos Totally agree, the large number of gas and fees is not a good user experience.  Let me think about how we can fix it later.

File: gno.land/pkg/sdk/vm/keeper.go
Comment (piux2) [reply gno.land/pkg/sdk/vm/keeper.go:208]: This is safe under the current ABCI protocol implementation. All ABCI requests to the app are synchronized at the ABCI protocol level. In other words, the VM Keeper processes one request at a time, which is why gnoStore can be shared.

Comment (piux2) [reply gno.land/pkg/sdk/vm/keeper.go:208]: This is safe under the current ABCI protocol implementation. All ABCI requests to the app are synchronized at the ABCI protocol level. In other words, the VM Keeper processes one request at a time, which is why gnoStore can be shared.

The GasMeter is created and wrapped in each transaction context before being set in the gnoStore.

Let me know if you still think this not safe.

File: gnovm/pkg/gnolang/store.go
Comment (piux2) [reply gnovm/pkg/gnolang/store.go:764]: > I think this ought to go in `BeginTransaction`, and the store shouldn't use a gas store otherwise.
@thehowl Agreed, I will assign the gasMeter in BeginTransaction and remove this method.

File: gno.land/pkg/sdk/vm/keeper.go
Comment (piux2) [reply gno.land/pkg/sdk/vm/keeper.go:208]: @thehowl  good catch! I will pass the gasMeter to gnoStore.BeginTransaction() instead

**Comment by Gno2D2** [comment]
I'm a bot that assists the Gno Core team in maintaining this repository. My role is to ensure that contributors understand and follow our guidelines, helping to streamline the development process.

The following requirements must be fulfilled before a pull request can be merged.
Some requirement checks are automated and can be verified by the CI, while others need manual verification by a staff member.

These requirements are defined in this [configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

## Automated Checks

 🟢 Maintainers must be able to edit this pull request
 🔴 The pull request head branch must be up-to-date with its base


## Manual Checks

*No manual checks match this pull request.*

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>The pull request head branch must be up-to-date with its base</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 Head branch (piux2:feat_gas_ops) is up to date with base (master): behind by 139 / ahead by 2

```
</blockquote></details>

</blockquote></details>



</blockquote></details>

File: gnovm/pkg/gnolang/store.go
Comment (piux2) [reply gnovm/pkg/gnolang/store.go:127]: I think it is better to keep the gas constant until we implement other ways to update the value, such as through genesis or GovDAO with the parameter module. For now, whenever we change the value, we should modify this section of the code and go through code reviews again.

File: tm2/pkg/sdk/context.go
Comment (piux2) [reply tm2/pkg/sdk/context.go:181]: > I can understand the rename to GasStore, but I don't think we should expose a gas-less `Store` method, especially when getting the underlying store is still trivial and exported.

The purpose here is not to rename the store but to separate gas store access from non-gas store access in the context. This separation is necessary because we need to wrap the gas meter at different levels for Auth Keeper gas consumption and VM gas consumption.

Please refer to the diagram above. In Auth Keeper, the GasStore wraps the gas meter at the underlying base store level. In VM Keeper, the gas meter is wrapped above the underlying store in the gnoStore, allowing us to capture activity to persist realm objects. The ctx.Store is used by the VM Keeper to access the underlying store without the gas meter, so it can wrap it at a higher layer.

The context.Store abstracts away the underlying MultiStore concept. This approach is consistent with how context, app, and keeper store access are managed.

File: gno.land/cmd/gnoland/testdata/issue_1786.txtar
Comment (piux2) [reply gno.land/cmd/gnoland/testdata/issue_1786.txtar:16]: Before this PR, the gas consumption for the VM was a placeholder. All opcodes consumed 1 gas, and the gas required to persist and load objects was not accounted for. It inherited the same gas meter consumption wrapped in the base store, which was used by the Auth Keeper and measured gas based on the raw byte size of each access.

File: gnovm/pkg/gnolang/machine.go
Comment (piux2) [reply gnovm/pkg/gnolang/machine.go:1003]: > Can we add a reference to the benchmarks and methodology?

I will write a separate article for the gas benchmarking.  It is related to PR https://github.com/gnolang/gno/pull/2241

**Pull request merged**
